### PR TITLE
Complete rewrite of the first tutorial

### PR DESCRIPTION
**Pull request by kevinpetit**
I have completely overhauled the first tutorial.

The tutorial has been broken down into 8 smaller chapters, where I have completely rewritten almost the entire tutorial.
Next to that, I've updated the code, indented it for readability and legibility, I have removed bad practices and direct styling, 
I've also added a bit about text sanitation in the last part which was absent in the older version of this tutorial - there's now no SQL-injection sensitive code left in the tutorial.

I'm going to rework the other tutorials too, and I'll modify the example files too (including a new Xoops Module icon).

Kevin

### CODE REVIEW

File: en/book/building_a_simple_module/saving_to_database.md [hunk 1]
```diff
@@ -12,12 +19,15 @@ if (isset($_POST['submit'])){
     if (empty($_POST['name'])){
```
Comment (geekwright) [inline en/book/building_a_simple_module/saving_to_database.md:19]: We should update this to match current best practices:
- avoid direct access to $_POST (of $_GET), use XoopsRequest instead. XoopsRequest abstracts the superglobal access, and provides sanitization of the input.
- addslashes should not be used for escaping strings for database, use $xoopsDB->escape() instead. There are potential character set related issues (some are exploitable) when using the addslashes() function. The escape() method of the database connector is reliable in all cases.

The result could look more like this:

```
if ('POST' === XoopsRequest::getMethod()){
    if (empty(XoopsRequest::getString('name', '')) {
        echo 'Please fill in a name';
    } else {
        //  You should NEVER trust anything that your users post - always first check it.
        // $_POST contains all the information from our form. We're creating a new variable to contain the submitted information.
        // This way, we can easily put them through to the database.
        $name       =   $xoopsDB->escape(XoopsRequest::getString('name'));
        $address    =   $xoopsDB->escape(XoopsRequest::getString('address', ''));
        $tel        =   $xoopsDB->escape(XoopsRequest::getString('tel', ''));
        $email      =   $xoopsDB->escape(XoopsRequest::getString('email', ''));
        // All values are done, let's prepare our SQL-query to store this information
        $query      =   "INSERT INTO ".$xoopsDB->prefix("tutorial_myform")." (name, address, telephone, email) VALUES ('$name', '$address', '$tel', '$email' )";
        // Let's perform the query
        $res        =   $xoopsDB->query($query);
        if (!$res) {
            echo "We've encountered an error: " . $xoopsDB->error;
        } else {
            echo "Data was correctly inserted into the database!";
        }
    }
}
```

**Comment by mambax7** [comment]
Thanks Richard for looking into it and providing feedback! This is very helpful!

**Comment by kevinpetit** [comment]
I'll update this - I suspected that there were better practices then those in the book.
My goal is to provide a complete modern rewrite for this, so please let me know if there are better ways to do things!

I'll also add some new tutorials once I'm done converting the existing ones.

**Comment by mambax7** [comment]
Kevin, you can also incorporate some of the best practices from here:
http://xoops.org/modules/newbb/viewtopic.php?post_id=353274

But they will also have to be reviewed by Richard to get a feedback that we're using practices that will position us well to move to XMF and later to XOOPS 2.6.0
There are always many ways to accomplish something, and while something might be technically correct and was used in XOOPS modules forever, we might use another approach in order to make it compatible with future coding standards of XMF and XOOPS 2.6.0

Thank you for doing it, and I hope, you're getting now more comfortable with GitBook! :)

**Comment by kevinpetit** [comment]
I'm currently rewriting the first chapter again, as I've decided to incorporate some things differently to make it user-friendlier. 
One of those is that the module is now going to be called "People", I've created a "blue-move" icon for the module, I've updated the xoops_info.php file with basename(**DIR**); so that it's cloneable and learns the best-practices from the start.

I'm currently using the https://github.com/XoopsModules25x/xmfdemo module as an example for me - I'm assuming that that is using the best practices there are right now.

One of the goals I have is once I'm done with writing the first few chapters, continue with writing my own chapters with XMF as the base.

I'm also thinking about renaming the book from Building a XOOPS Module by Kaotik to **Building XOOPS modules**, as it's pretty much a complete rewrite and the scope is going to be completely different.

I do hope that Richard is willing to review all the documentation I'm writing (or at least the code bits), so that we have some really great documentation available (so that we might get some more developers! :)).

**Comment by kevinpetit** [comment]
Michael, Richard,

How can I make the module cloneable with the SQL? I'm not really sure how to do that.
Everything else is cloneable, but afaik I still have to hardcode a name in the SQL file. Is there a solution for this?

Kev

**Comment by mambax7** [comment]
You might look at the D3 modules from GIJoe (e.g. [Pico](https://github.com/XoopsX/pico)). There is [an intro here](http://www.mdxprod.com/modules/news/article.php?storyid=153) with some extra links to GIJoe's Website

**Comment by geekwright** [comment]
> I do hope that Richard is willing to review all the documentation I'm writing (or at least the code bits), so that we have some really great documentation available (so that we might get some more developers! :)).

Count me in! Sounds like we share the same goal. Thanks for all your efforts!

> How can I make the module cloneable with the SQL?

Publisher has a pretty clean clone implementation in admin/clone.php. It handles the sql table name prefixes. Essentially, it starts with the current dirname and replaces 'dirname', 'DIRNAME' and 'Dirname' in each file, writing the changes to the new directory.

To me, it seems a little easier to deal with than D3, but they both have their strengths.

Just looking to the future, PHP namespaces will be coming to the module specification in the next generation, and that will change things like the magic function names with the embedded dirnames like  "xoops_module_install__dirname_". We'll shift to specific methods in specific classes within the namespace for the module<->core interactions. This is just a heads up that we will revisit cloning in earnest once that is in place.

**Comment by kevinpetit** [comment]
Richard/Michael, how do I go around with including jQuery on the front-end of a module? 
I would prefer to use the version already in XOOPS and try to avoid duplication, but I can't really find info that's helping me.

How can I do this best?

BTW: I need a better way than this to ask questions :P I'll have a lot of them.

**Comment by mambax7** [comment]
1) Asking questions here is fine - just group similar questions together and open a new issue here

2) For jQuery, you can reuse the one from XOOPS this way:

```
    /** add JQuery and JQuery UI */
    global $xoTheme;
    $xoTheme->addScript("browse.php?Frameworks/jquery/jquery.js");
    $xoTheme->addScript("browse.php?Frameworks/jquery/plugins/jquery.ui.js");
```

**Comment by geekwright** [comment]
Hi @kevinpetit, things are looking good!

A little observation on the memberlist.md section. I'm looking at this section, in particular, but the comments apply to all the similar code. You have the following code:

``` PHP
    // Load up the XOOPS member handler
    $member_handler =& xoops_gethandler('member');
    // Grab all our members
    $allUsers =& $member_handler->getUsers();
    // Now let's loop through all our users
    foreach(array_keys($allUsers) as $i) {
        var_dump($allUsers[$i]);
        echo '<br>';
    }
```

I would suggest replacing that with this:

``` PHP
    // Load up the XOOPS member handler
    $memberHandler = xoops_getHandler('member');
    // Grab all our members
    $allUsers = $memberHandler->getUsers();
    // Now let's loop through all our users
    foreach ($allUsers as $user) {
        \Xmf\Debug::dump($user);
    }
```
- we should be moving to camel case variable names as possible, so `$member_handler` becomes `$memberHandler`
- xoops_getHandler() is the actual name so we should match the case. (The name is bad style, but we need to keep it for compatibility for now.)
- xoops_getHandler() returns an object, so the `=&` to assign by reference is a strict warning in PHP5 and an error in PHP7. Lots of code has this pattern, but it has roots in PHP 4. Michael and I have spent many hours removing that pattern in XOOPS 2.5.8 :grinning: 
- the `->getUsers();` call returns an array, so the `=&` is not technically an error, but again the pattern goes back to PHP 4 where it saved some memory. With the introduction of "copy on write" memory management in PHP 5, it really isn't needed any more.
- Eliminating the use of the array index variables lets `foreach` loop  directly through the array objects, which is cleaner and more efficient.
- XOOPS 2.5.8 introduced some sweet debugging tools, so can we consider swapping in Xmf\Debug::dump() for var_dump()? It is always there, and always works the same, so learning to use it as part of XOOPS module coding sounds like a good idea.

I know this sounds like a lot nit picking, but that is just a sign of how important it is to update this tutorial! PHP has matured, standards, such as PSR-2 have become the norm. We have been updating core and modules to reflect the modern standards, so it will be great to have tutorials which respect those standards.

Thanks again for all the improvements!

**Comment by kevinpetit** [comment]
I don't see this as nit picking Richard, far from it. This documentation needs to be PERFECT, it needs to be THE resource for XOOPS module development. So it should use best practices.

That's why I have decided to only use the modern techniques - all styling that I'll teach will apply to xbootstrap (or just regular Bootstrap) themes. I won't learn how to theme for the older legacy themes (though I might add this in a new chapter later on). It's important to look forward now, not backwards.

So please, let me know if ANYTHING at all is wrong or could be done better - the code changes that you've suggested made the code a lot easier and clear to work with too ;)!

BTW: the next tutorial goes about AJAX, which is not exactly my cup of tea, but I'll try to do this. What is the best way to incorporate AJAX in XOOPS? Kaotik used xajax for this, but it looks like that's a dead project. No use investing time to teach a dead product..

Kevin

**Comment by geekwright** [comment]
:+1: I'll keep reviewing things. We seem to be clearly in alignment on the need for quality!

For ajax, the PHP side is fairly simple, and jquery should be the standard on the front end.

There is a jwtdemo.php in the xmfdemo module that could be used as a basis. My big concern with ajax is security. A lot of production code has at least the potential of being abused. Stressing JWT as part of a secure solution should be part of the tutorial.

There is some documentation on JWT in the [xmf-cookbook](https://xoops.gitbooks.io/xmf-cookbook/content/en/book/recipes/jsonwebtokens.html). If you have questions, just let me know. Feel free to drop me direct email.

Thanks for all the good work!

**Comment by kevinpetit** [comment]
Do you mean the jwtajax.php file?

**Comment by geekwright** [comment]
Opps. Yes, that would be the one.

**Comment by kevinpetit** [comment]
I've started working on the module - a demo of it (it's still rough!) can be found on http://testrealm.vinpet.it/xoops/modules/people/ - let me know what you think.
I'll put the code up as soon as I can.

I'm not really sure how to do it in the proper way, I'm just doing it like I would do it in my own projects (without XOOPS). Everything AJAX/JavaScript has been implemented by using jQuery.

I'm not loading jQuery myself in the module, I'm relying on the theme to do that. Is there a reliable way to check that it's loaded?

What I have done is I have created a new file, called includes/functions.php and ajax.php. In the functions.php I basicly put the function that gets the data of users:

``` php
<?php
/**
 * Created by PhpStorm.
 * User: Petit_Kevin
 * Date: 19/07/2016
 * Time: 15:08
 */

function dbGetAllPeopleData() {
    // Get the global xoopsDB
    global $xoopsDB;
    // Create a new empty array client
    $person = array();
    // A nice counter set to 1
    $i = 1;
    // Let's create our query.
    $query = $xoopsDB->query('SELECT * FROM '. $xoopsDB->prefix('people_myform'));
    // Perform our query and loop through it
    while($myRow = $xoopsDB->fetchArray($query)) {
        // First run, this will be $client[1]['name'], next one $client[2]['name'], etc.
        $person[$i]['name']         =   $myRow['name'];
        $person[$i]['address']      =   $myRow['address'];
        $person[$i]['telephone']    =   $myRow['telephone'];
        $person[$i]['email']        =   $myRow['email'];
        // Let's add one to the counter
        $i++;
    }
    // Let's return our array.
    return $person;
}
```

In the ajax.php file I created 2 if statements, which will return the information in a table or load up the form - this is obviously deviating FAR from how I should do it, but this does work (except for the translation information). I still have to leverage using the smarty templates for this.

``` php
<?php
/**
 * Created by PhpStorm.
 * User: Petit_Kevin
 * Date: 19/07/2016
 * Time: 13:59
 */
require_once dirname(dirname(__DIR__)) . '/mainfile.php';

if(!empty(Xmf\Request::getString('listAll', ''))) {
    echo '<table class="table table-striped" width="100" border="0">
                    <tr>
                        <td>'.PP_NAME.'</td>
                        <td>'.PP_ADDRESS.'</td>
                        <td>'.PP_TELEPHONE.'</td>
                        <td>'.PP_EMAIL.'</td>
                    </tr>';
    $query = $xoopsDB->query('SELECT * FROM '. $xoopsDB->prefix(basename(__DIR__)."_myform"));
    while($myRow = $xoopsDB->fetchArray($query)) {
        $name       = $myRow['name'];
        $address    = $myRow['address'];
        $telephone  = $myRow['telephone'];
        $email      = $myRow['email'];
        echo '<tr><td>'.$name.'</td><td>'.$address.'</td><td>'.$telephone.'</td><td>'.$email.'</td></tr>';
    }
    echo '</table>';
}

if(!empty(Xmf\Request::getString('addNew', ''))) {
    echo '<div id="empty"></div>
<form class="form" name="people_form" method="post" id="people_form" action="index.php">
    <fieldset class="form-group">
    <div class="form-group">
        <label for="name"><{$smarty.const.PP_NAME}></label>
        <input type="text" class="form-control" name="name" placeholder="<{$smarty.const.PP_NAME}>">
    </div>
    <div class="form-group">
        <label for="name"><{$smarty.const.PP_ADDRESS}></label>
        <input type="text" class="form-control" name="address" placeholder="<{$smarty.const.PP_ADDRESS}>">
    </div>
    <div class="form-group">
        <label for="name"><{$smarty.const.PP_TELEPHONE}></label>
        <input type="text" class="form-control" name="telephone" placeholder="<{$smarty.const.PP_TELEPHONE}>">
    </div>
    <div class="form-group">
        <label for="name"><{$smarty.const.PP_EMAIL}></label>
        <input type="text" class="form-control" name="email" placeholder="<{$smarty.const.PP_EMAIL}>">
    </div>
    <button class="btn btn-default" type="submit" name="listAll" value="List All" id="listAll">List all inputs</button>
    <button class="btn btn-success" type="submit" name="submit" value="submit">Submit</button>
    </fieldset>
</form>
';
}
```

I realize why this is bad - it's basicly repeating myself. I think the best way to do things is to let the javascript call the ajax.php file and let it return the raw HTML, provided by the Smarty templates.

What do you think about this? I'm going to add te code of the main template (with all the javascript code in it) as well:

``` php
<{$msg}>
<{if $addNew==0}>
    <table width="100%" border="0" cellspacing="0" cellpadding="0" id="peopleAddNewButton">
        <tr>
            <td><a class="btn btn-default" id="addNew" href="index.php?addNew=1">Add new person</a></td>
        </tr>
    </table>
<{/if}>
<{if $addNew==1}>
    <{include file="db:people_form.html"}>
<{/if}>
<div id="empty"></div>
<table class="table table-striped" width="100%" border="0" cellspacing="0" cellpadding="0" id="peopleTable">
    <tr>
        <td><{$smarty.const.PP_NAME}></td>
        <td><{$smarty.const.PP_ADDRESS}></td>
        <td><{$smarty.const.PP_TELEPHONE}></td>
        <td><{$smarty.const.PP_EMAIL}></td>
    </tr>
    <{foreach item = person from = $persons}>
    <tr>
        <td><{$person.name}></td>
        <td><{$person.address}></td>
        <td><{$person.telephone}></td>
        <td><{$person.email}></td>
    </tr>
    <{/foreach}>
</table>
<script>
    $(document).ready(function()
    {
        $("#people_form").submit(function(e)
        {
            if($('input:text').is(":empty")) {
                $("#empty").html("<p class='bg-danger'><{$smarty.const.PP_ENTERNAME}></p>");
                e.preventDefault();
            } else {
                e.preventDefault();
                $.ajax({
                    url : $(this).attr('action') || window.location.pathname,
                    type: 'post',
                    data: $(this).serialize(),
                    success: function (data) {
                        $("#empty").html("Person added!");
                    },
                    error: function (jXHR, textStatus, errorThrown) {
                        alert(errorThrown);
                    }
                });
            }
        });

        $("#listAll").click(function(e)
        {
            e.preventDefault();
            $("#empty").load("ajax.php?listAll=1");

        });

        $("#addNew").click(function(e)
        {
            e.preventDefault();
            $("#empty").load("ajax.php?addNew=1");
            $("#peopleTable").fadeOut();
            $("#addNew").hide();
        })
    });
</script>
```

I realize I'm probably going about this completely the wrong way, but I wanted to try to find "what worked". I hope you guys will be able to tell me what the best way is to implement this - I can't imagine that XOOPS hasn't got functions for this?

Kev

**Comment by kevinpetit** [comment]
Richard/Michael, have you got any time to look at this?

**Comment by mambax7** [comment]
It's looking good! 
Suggestion: move the functions file to /class and rename it to "utilities.php", and make it a class with static functions. This way you can have it more OOP, but also the developers know where each function is coming from, and the functions don't clutter the global space.  
I'm looking forward to the next part...

**Comment by geekwright** [comment]
Kevin,

First, I am incredibly sorry about the delay. Please accept my sincere apology.

I'm going to make a few suggestions, and they are going to come in stages, so you can see how these relate, and better understand the reasons.

Let's consider this piece:

``` PHP
function dbGetAllPeopleData() {
    // Get the global xoopsDB
    global $xoopsDB;
    // Create a new empty array client
    $person = array();
    // A nice counter set to 1
    $i = 1;
    // Let's create our query.
    $query = $xoopsDB->query('SELECT * FROM '. $xoopsDB->prefix('people_myform'));
    // Perform our query and loop through it
    while($myRow = $xoopsDB->fetchArray($query)) {
        // First run, this will be $client[1]['name'], next one $client[2]['name'], etc.
        $person[$i]['name']         =   $myRow['name'];
        $person[$i]['address']      =   $myRow['address'];
        $person[$i]['telephone']    =   $myRow['telephone'];
        $person[$i]['email']        =   $myRow['email'];
        // Let's add one to the counter
        $i++;
    }
    // Let's return our array.
    return $person;
}
```

Every time you add or change a field in the database, you will have to touch this code. If, for example, you add an 'verified' column, you will have to add it to your list of assignments.

What if instead we did this:

``` PHP
function dbGetAllPeopleData() {
    // Get the global xoopsDB
    global $xoopsDB;
    // Create a new empty array client
    $person = array();

    // Let's create our query.
    $query = $xoopsDB->query('SELECT * FROM '. $xoopsDB->prefix('people_myform'));
    // Perform our query and loop through it
    while($myRow = $xoopsDB->fetchArray($query)) {
        $person[] = $myRow;
    }
    // Let's return our array.
    return $person;
}
```

Short and sweet. We don't need to keep track of the array index, just add to the end, so we have one less thing to code. We get an array from fetchArray(), so we can just add it directly, and all the fields in the query go with it. We can sort out what we want to show in the presentation layer, instead of embedding it in some lower level code. Big wins all around.

Can we do better?

If we back up a step, how about instead of using SQL, we went with the XoopsPersitableObjectHandler instead? There are times when SQL is a must, for example a complex join or condition the objects don't handle well. There are a world of advantages that can simplify the code with reusable  methods replacing copy and paste repetition.

(Full disclosure, I am guilty of being a terrible example in this area, but  I am trying to do better.)

First we need two objects, a XoopsObject to describe our data, and a Handler that will define the basics, like the table name, the primary key, and the data object.

Create this as a file, class/person.php

```
class PeoplePerson extends XoopsObject
{
    public function __construct()
    {
        parent::__construct();
        $this->initVar('id', XOBJ_DTYPE_INT);
        $this->initVar('name', XOBJ_DTYPE_TXTBOX);
        $this->initVar('address', XOBJ_DTYPE_TXTAREA);
        $this->initVar('telephone', XOBJ_DTYPE_TXTBOX);
        $this->initVar('email', XOBJ_DTYPE_EMAIL);
    }
}

class PeoplePersonHandler extends XoopsPersistableObjectHandler
{
    public function __construct(XoopsDatabase $db)
    {
        parent::__construct($db, 'people_person', 'PeoplePerson', 'id', 'name');
    }
}
```

We describe the table columns in PeoplePerson, and the table, data object, primary key, and a descriptive column. Not much so far, but consider this bit of code:

``` PHP
function dbGetAllPeopleData() {
    $helper = \Xmf\Module\Helper::getHelper('people');
    $handler = $helper->getHandler('person');

    $criteria = new CriteriaCompo();
    $criteria->setSort('name');
    $criteria->setOrder('DESC');

    $person = $handler->getAll($criteria, null, false, false);

    // Let's return our array.
    return $person;
}
```

What did we gain? Lots!

We are letting XOOPS do what it does best, taking care of the details. We don't need the $xoopsDD global, we are not concerned with prefixes, we gain all  sorts of control with the criteria classes, and we gain a lot of portability by reusing the system code.

I'll post this part, and work on the next part -- it might be tomorrow before I can get it, but it is coming.

Again, sorry for being inattentive.

_(edited to correct issue in address datatype, thanks @zyspec )_

**Comment by zyspec** [comment]
Kevin/Richard,

There's one slight change to Richard's implementation.  In the 
dbGetAllPeopleData::__construct() method need to change the 'address' 
init from:

|$this->initVar('address', XOBJ_DTYPE_INT); |

  to

|$this->initVar('address', XOBJ_DTYPE_TXTBOX);|

**Comment by geekwright** [comment]
Doh! It was intended to be a XOBJ_DTYPE_TXTAREA.

**Comment by geekwright** [comment]
> I'm not loading jQuery myself in the module, I'm relying on the theme to do that. Is there a reliable way to check that it's loaded?

Unfortunately, the only way right now is to check in the browser. (This was one of many motivations for the asset management in 2.6.) [Here](http://jquery-howto.blogspot.com/2009/03/check-if-jqueryjs-is-loaded.html) is some discussion on how to do this client side.

**Comment by geekwright** [comment]
This continues on uses for the PeoplePerson XoopsObject discussed previously.

Once you have the data nicely encapsulated in a XoopsObject, it becomes easier to get the data where it needs to be used. 

Consider building the form used to edit the data. XOOPS has form building tools that take care of the details of building forms and form elements. In addition to displaying a complete form, the XoopsForm objects support assigning the form details to a Smarty variable, which can be used to gain extra control over the form layout, making forms which use them customizable with theme template overrides.

It would be useful to be able to code a form once, and then use it in several places, for example on both the user and administration sides. To do that, we need to move our form building into a function. Once we do that, we need to pass our data into it. 

Without an object the function might end up with a signature like this:

``` PHP
function buildForm($name, $address, $telephone, $email)
{
    // ... 
    $form->addElement(new XoopsFormText(_MA_PEOPLE_PERSON_NAME, 'name', 50, 255, $name), true);
    $form->addElement(new XoopsFormTextArea(_MA_PEOPLE_PERSON_ADDRESS, 'address', $address));
    $form->addElement(new XoopsFormText(_MA_PEOPLE_PERSON_TELEPHONE, 'telephone', 32, 32, $telephone));
    $form->addElement(new XoopsFormText(_MA_PEOPLE_PERSON_EMAIL, 'email', 50, 255, $email));
    // ...
}
```

That creates the same sort of maintenance headache we mentioned before --  add a field and you end up changing things spread all over the place. 

With a data object, we can simplify that signature:

``` PHP
function buildForm(PeoplePersonObject $peoplePerson)
{
    // ...
    $form->addElement(new XoopsFormText(_MA_PEOPLE_PERSON_NAME, 'name', 50, 255, $dataObject->getVar('name')), true);
    $form->addElement(new XoopsFormTextArea(_MA_PEOPLE_PERSON_ADDRESS, 'address', $dataObject->getVar('address')));
    $form->addElement(new XoopsFormText(_MA_PEOPLE_PERSON_TELEPHONE, 'telephone', 32, 32, $dataObject->getVar('telephone')));
    $form->addElement(new XoopsFormText(_MA_PEOPLE_PERSON_EMAIL, 'email', 50, 255, $dataObject->getVar('email')));
    // ...
}
```

Instead of a simple function, how about using a class instead? Consider for a minute that we just talked a function named "buildForm". That has a huge chance of collision. Even if it is unique in your module, there could be other modules installed. Those modules might have blocks which will be displayed at the same time your module is in use, and those blocks might include their own functions.php style file -- chaos ensues. This was why Michael suggested a class of static methods earlier.

How about we take that one step further, and build a class just for our form? Come to think of it, how
about just extending a XoopsForm class? Actually, that is exactly what the newest modules, like publisher, do.

This could be class/form/person.php

``` PHP
<?php

use Xmf\Request;

xoops_load('XoopsFormLoader');
xoops_loadLanguage('main', 'people');

class PeoplePersonForm extends XoopsThemeForm
{
    public function __construct(XoopsObject $dataObject, $name = 'personform', $action = '', $method = 'post')
    {
        parent::__construct(_MA_PEOPLE_PERSON_FORM, $name, $action, $method, true);

        static::createElements($this, $dataObject);
        static::createButtons($this, $dataObject);
    }

    public static function createElements(XoopsForm $form, XoopsObject $dataObject)
    {
        $name = new XoopsFormText(_MA_PEOPLE_PERSON_NAME, 'name', 50, 255, $dataObject->getVar('name', 'e'));
        $form->addElement($name, true);

        $address = new XoopsFormTextArea(_MA_PEOPLE_PERSON_ADDRESS, 'address', $dataObject->getVar('address', 'e'));
        $form->addElement($address);

        $telephone = new XoopsFormText(_MA_PEOPLE_PERSON_TELEPHONE, 'telephone', 32, 32, $dataObject->getVar('telephone', 'e'));
        $form->addElement($telephone);

        $email = new XoopsFormText(_MA_PEOPLE_PERSON_EMAIL, 'email', 50, 255, $dataObject->getVar('email', 'e'));
        $form->addElement($email, true);

        $form->addElement(new XoopsFormHidden('id', $dataObject->getVar('id')));
    }

    public static function createButtons(XoopsForm $form, XoopsObject $dataObject)
    {
        $button = new XoopsFormButton('', 'submit', _MA_PEOPLE_FORM_SUBMIT, 'submit');
        $form->addElement($button);
    }

    public static function applyFormInput(XoopsObject $dataObject)
    {
        $dataObject->setVar('name', Request::getString('name', $dataObject->getVar('name', 'e'), 'POST'));
        $dataObject->setVar('address', Request::getString('address', $dataObject->getVar('address', 'e'), 'POST'));
        $dataObject->setVar('telephone', Request::getString('telephone', $dataObject->getVar('telephone', 'e'), 'POST'));
        $dataObject->setVar('email', Request::getEmail('email', $dataObject->getVar('email', 'e'), 'POST'));
    }
}
```

In addition to putting object data info the form, we've add a method to move form data to the object, keeping all of that form related data movement in one place.

We can put this class to work in many ways. This shell shows some ideas:

`````` PHP
<?php

use Xmf\Request;

require_once dirname(dirname(__DIR__)) . '/mainfile.php';

include_once __DIR__ . '/class/form/person.php';

$xoopsOption['template_main'] = 'db:people_personform.tpl';
require_once XOOPS_ROOT_PATH . '/header.php';

$helper = \Xmf\Module\Helper::getHelper('people');
/** @var PeoplePersonHandler $handler */
$handler = $helper->getHandler('person');

$select_id = Request::getInt('select_id', '0', 'GET');
$select_id = Request::getInt('id', $select_id, 'POST');
$personObject = $handler->get($select_id);
if (empty($personObject)) {
    $personObject = $handler->create();
}

PeoplePersonForm::applyFormInput($personObject);

if ('POST' === Request::getMethod()) {
    $validated = true;
    $message = '';

    // should validate the data here

    if (!$GLOBALS['xoopsSecurity']->validateToken()) {
        $validated = false;
        // token error
        $message .= $GLOBALS['xoopsSecurity']->getErrors(true);
    }

    if ($validated) {
        $status = $handler->insert($personObject);
        $errors = $handler->getErrors();
        if (!empty($errors)) {
            $message .= implode('<br>', $error);
        }
        if ($status !== false) {
            $message .= 'Updated.';
        }
    }
    $GLOBALS['xoopsTpl']->assign('message', $message);
    $GLOBALS['xoopsTpl']->assign('validated', $validated);
}

$form = new PeoplePersonForm($personObject, 'personform', basename(__FILE__));
Debug::dump($_POST, Request::getMethod(), $personObject, $form);
$form->assign($GLOBALS['xoopsTpl']);

$listForm = new XoopsForm('Select a Person', 'personpicker', basename(__FILE__), 'GET');
$list = $handler->getList();
$list += array('0' => '(new)');
$selectList = new XoopsFormSelect('People', 'select_id', $select_id);
$selectList->addOptionArray($list);
$selectList->setExtra(' onchange="document.getElementById(\'personpicker\').submit();" ');
$listForm->addElement($selectList);
$listForm->addElement(new XoopsFormButton('', 'ppsubmit', 'Submit', 'submit'));
$listForm->assign($GLOBALS['xoopsTpl']);

require_once XOOPS_ROOT_PATH . '/footer.php';```

Rather than directly display the forms, we assign them to the template so we can cutomize them. This is an example template, that works well in bootstrap themes:
```HTML
<div class="people-module">
    <{if $message|default:false}>
        <{if $validated|default:false}>
            <div class="alert alert-success alert-dismissible" role="alert">
        <{else}>
            <div class="alert alert-danger alert-dismissible" role="alert">
        <{/if}>
        <button type="button" class="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button>
        <{$message}>
        </div>
    <{/if}>
    <{$personform.javascript}>
    <form name="<{$personform.name}>" action="<{$personform.action}>" method="<{$personform.method}>"
    <{$personform.extra}>>
    <h4><{$personform.title}></h4>
    <{foreach item=element from=$personform.elements}>
    <{if $element.hidden != true}>
    <div class="form-group">
        <label><{$element.caption}></label>
        <{$element.body|replace:'formButton':'btn btn-default'|replace:' id=':' class="form-control" id='}>
    </div>
    <{else}>
    <{$element.body}>
    <{/if}>
    <{/foreach}>
    </form>
                <{$personpicker.javascript}>
    <form id="<{$personpicker.name}>" name="<{$personpicker.name}>" action="<{$personpicker.action}>" method="<{$personpicker.method}>" <{$personpicker.extra}>>
    <h4><{$personpicker.title}></h4>
    <{foreach item=element from=$personpicker.elements}>
    <{if $element.hidden != true}>
    <div class="form-group">
        <label><{$element.caption}></label>
        <{$element.body|replace:'formButton':'btn btn-default'|replace:' id=':' class="form-control" id='}>
    </div>
    <{else}>
    <{$element.body}>
    <{/if}>
    <{/foreach}>
    </form>
</div>
``````

I hope this gives you some answers and some ideas.

**Pull request merged**